File: bindings/python/bro.cpp

#include <utility>
#include <string>

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
#include <pybind11/pybind11.h>
#pragma GCC diagnostic pop

#include "broker/bro.hh"
#include "broker/data.hh"

namespace py = pybind11;
using namespace pybind11::literals;

void init_bro(py::module& m) {
  py::class_<broker::bro::Message>(m, "Message")
    .def("as_data", &broker::bro::Message::as_data);

  py::class_<broker::bro::Event, broker::bro::Message>(m, "Event")
    .def(py::init([](broker::data data) {
       return broker::bro::Event(std::move(data));
       }))
    .def(py::init([](std::string name, broker::data args) {
       return broker::bro::Event(std::move(name), std::move(caf::get<broker::vector>(args)));
       }))
    .def("name",
          static_cast<const std::string& (broker::bro::Event::*)() const>
          (&broker::bro::Event::name))
    .def("args",
         static_cast<const broker::vector& (broker::bro::Event::*)() const>
         (&broker::bro::Event::args));
}

File: bindings/python/broker/bro.py

try:
    from . import _broker
except ImportError:
    import _broker

import broker

class Event(_broker.bro.Event):
    def __init__(self, *args):
        if len(args) == 1 and isinstance(args[0], list):
            # Parse raw broker message as event.
            _broker.bro.Event.__init__(self, broker.Data.from_py(args[0]))
        elif len(args) == 2 and isinstance(args[0], str) and not isinstance(args[0], broker.Data):
            # (name, (arg1, arg2, ...))
            _broker.bro.Event.__init__(self, args[0], broker.Data.from_py(args[1]))
        else:
            # (name, arg1, arg2, ...)
            _broker.bro.Event.__init__(self, args[0], broker.Data.from_py(args[1:]))

    def args(self):
        return [broker.Data.to_py(a) for a in _broker.bro.Event.args(self)]
